qiime2/core/annotate.py

```diff
@@ -351,3 +395,319 @@ def _write(self, annotations_dir, root_result_uuid,
+    if algorithm == 'EdDSA' and curve == 'ed25519':
+        return 'Ed25519'
+    elif algorithm in {'ECDSA', 'ECDH'} and key_info.get('curve'):
+        return f'{algorithm}/{key_info["curve"]}'
+    elif algorithm in {'RSA', 'DSA'} and length:
+        return f'{algorithm}-{length}'
+    else:
+        return algorithm or 'unknown'
+
+
+class Signature(Annotation):
+    """Signature sub-class, inherits from Annotations.
+
+    Creates a cryptographic signature over the Result's root checksums.sha512
+    file using credentials for an existing key pair via GnuPG.
+
+    Parameters
+    ----------
+    name : str
+        Annotation name (validated like other Annotations).
+
+    signer_name : str, optional
+        Name associated with the key pair in GnuPG
+        that will be used for signing.
+        At least one of name/email must be provided for Signature creation.
+
+    signer_email : str, optional
+        Email associated with the key pair in GnuPG
+        that will be used for signing.
+        At least one of name/email must be provided for Signature creation.
+
+    Returns
+    -------
+    Signature : obj
+        The instantiated Signature.
+
+    See Also
+    --------
+    Annotation
+    """
+    annotation_type = 'Signature'
+
+    # NOTE: in future versions, name will become optional & the default value
+    # will be the annotation's UUID (if name isn't provided by the user)
+    def __init__(self, name, *, signer_uid=None, fingerprint=None):
+        self.validate_name(name)
+
+        # Ensure at least one of signer uid (name/email address) is provided
+        if not signer_uid and not fingerprint:
+            raise ValueError(
+                'No inputs provided to either `signer_uid` or `fingerprint`. '
+                'Please provide either signer ID (name and email) or '
+                'fingerprint for key pair identification.'
+            )
+
+        # Construct Annotation class
+        self.signer_uid = signer_uid
+        self.fingerprint = fingerprint
+        super().__init__(name)
+
+    def _write(self, annotations_dir, root_result_uuid,
+               referenced_result_uuid):
+        """Write the contents of an instantiated Signature.
+
+        Parameters
+        ----------
+        annotations_dir
+            The path to the `annotations` directory within a Result object.
+            Located under `provenance`.
+
+        root_result_uuid
+            The uuid of the Result object where an Annotation is being added.
+
+        referenced_result_uuid
+            The uuid of the Result object that an Annotation is referring to.
+            Note that in 7.0 & 7.1, `root_result_uuid` and
+            `referenced_result_uuid` are the same (i.e. Annotations can only
+            refer to the same Result they are being attached to) but separate
+            root and referenced uuids will be supported in future versions.
+        """
+        root_fp = _find_root_fp(annotations_dir)
```

Comment:
```suggestion
        root_fp = _find_root_fp(annotations_dir, root_result_uuid)
```

Only if `_find_root_fp` is changed